workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowUiViewState.kt

```diff
@@ -0,0 +1,192 @@
+ */
+@WorkflowUiExperimentalApi
+public fun <RenderingT : Any> View.showRendering(
+  rendering: RenderingT,
+  viewEnvironment: ViewEnvironment
+) {
+  workflowTagAsStarted.let { tag ->
+    check(tag.showing.matches(rendering)) {
+      "Expected $this to be able to show rendering $rendering, but that did not match " +
+        "previous rendering ${tag.showing}. " +
+        "Consider using WorkflowViewStub to display arbitrary types."
+    }
+
+    // Update the tag's rendering and viewEnvironment.
+    workflowTag = Started(rendering, viewEnvironment, tag.showRendering)
+    // And do the actual showRendering work.
+    tag.showRendering.invoke(rendering, viewEnvironment)
+  }
+}
+
+/**
+ * Returns the most recent rendering shown by this view cast to [RenderingT],
+ * or null if [bindShowRendering] has never been called.
+ *
+ * @throws ClassCastException if the current rendering is not of type [RenderingT]
+ */
+@WorkflowUiExperimentalApi
+public inline fun <reified RenderingT : Any> View.getRendering(): RenderingT? {
+  // Can't use a val because of the parameter type.
+  return when (val showing = workflowTagOrNull?.showing) {
+    null -> null
+    else -> showing as RenderingT
+  }
+}
+
+/**
+ * Returns the most recent [ViewEnvironment] applied to this view, or null if [bindShowRendering]
+ * has never been called.
+ */
+@WorkflowUiExperimentalApi
+public val View.environment: ViewEnvironment?
+  get() = workflowTagOrNull?.environment
+
+/**
+ * Returns the function set by the most recent call to [bindShowRendering], or null
+ * if that method has never been called.
+ */
+@WorkflowUiExperimentalApi
+public fun <RenderingT : Any> View.getShowRendering(): ViewShowRendering<RenderingT>? {
+  return workflowTagOrNull?.showRendering
+}
+
+@WorkflowUiExperimentalApi
+internal var View.starter: (View) -> Unit
+  get() = workflowTagAsNew.starter
+  set(value) {
+    workflowTag = workflowTagAsNew.copy(starter = value)
+  }
+
+@WorkflowUiExperimentalApi
+@PublishedApi
+internal val View.workflowTagOrNull: WorkflowUiViewState<*>?
+  get() = getTag(R.id.workflow_ui_view_state) as? WorkflowUiViewState<*>
+
+@WorkflowUiExperimentalApi
+private var View.workflowTag: WorkflowUiViewState<*>
+  get() = workflowTagOrNull ?: error(
+    "Expected $this to have been built by a ViewFactory. " +
+      "Perhaps the factory did not call View.bindShowRendering."
+  )
+  set(value) = setTag(R.id.workflow_ui_view_state, value)
+
+@WorkflowUiExperimentalApi
+private val View.workflowTagAsNew: New<*>
+  get() = workflowTag as? New<*> ?: error(
+    "Expected $this to be un-started, but View.start() has been called"
+  )
+
+@WorkflowUiExperimentalApi
+private val View.workflowTagAsStarted: Started<*>
+  get() = workflowTag as? Started<*> ?: error(
+    "Expected $this to have been started, but View.start() has not been called"
+  )
+
+@WorkflowUiExperimentalApi
+private fun Any.matches(other: Any) = compatible(this, other)
```

Comment:
I get that maybe this makes it more fluent in `showRendering` above but adding another layer of indirection is a lot to carry in your mind when grokking.

Not a big deal but maybe can we just use `compatible(tag.showing, rendering)` above? Is there another use for this?